internal/nfs/nodeserver/nodeserver.go

```diff
@@ -0,0 +1,346 @@
+	ctx context.Context,
+	req *csi.NodeExpandVolumeRequest,
+) (*csi.NodeExpandVolumeResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+// NodeGetCapabilities returns the supported capabilities of the node server.
+func (ns *NodeServer) NodeGetCapabilities(
+	ctx context.Context,
+	req *csi.NodeGetCapabilitiesRequest,
+) (*csi.NodeGetCapabilitiesResponse, error) {
+	return &csi.NodeGetCapabilitiesResponse{
+		Capabilities: []*csi.NodeServiceCapability{
+			{
+				Type: &csi.NodeServiceCapability_Rpc{
+					Rpc: &csi.NodeServiceCapability_RPC{
+						Type: csi.NodeServiceCapability_RPC_GET_VOLUME_STATS,
+					},
+				},
+			},
+			{
+				Type: &csi.NodeServiceCapability_Rpc{
+					Rpc: &csi.NodeServiceCapability_RPC{
+						Type: csi.NodeServiceCapability_RPC_SINGLE_NODE_MULTI_WRITER,
+					},
+				},
+			},
+		},
+	}, nil
+}
+
+// NodeGetVolumeStats get volume stats.
+func (ns *NodeServer) NodeGetVolumeStats(
+	ctx context.Context,
+	req *csi.NodeGetVolumeStatsRequest,
+) (*csi.NodeGetVolumeStatsResponse, error) {
+	var err error
+	targetPath := req.GetVolumePath()
+	if targetPath == "" {
+		return nil, status.Error(codes.InvalidArgument,
+			fmt.Sprintf("targetpath %v is empty", targetPath))
+	}
+
+	stat, err := os.Stat(targetPath)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument,
+			"failed to get stat for targetpath %q: %v", targetPath, err)
+	}
+
+	if stat.Mode().IsDir() {
+		return csicommon.FilesystemNodeGetVolumeStats(ctx, ns.Mounter, targetPath)
+	}
+
+	return nil, status.Errorf(codes.InvalidArgument,
+		"targetpath %q is not a directory or device", targetPath)
+}
+
+// mountNFS mounts nfs volumes.
+func (ns *NodeServer) mountNFS(
+	ctx context.Context,
+	volumeID, clusterID, source, mountPoint string,
+	mountOptions []string,
+	mountPermissions uint64,
+) error {
+	var (
+		stderr, netNamespaceFilePath string
+		err                          error
+	)
+
+	if clusterID != "" {
+		netNamespaceFilePath, err = util.GetNFSNetNamespaceFilePath(
+			util.CsiConfigFile,
+			clusterID)
+		if err != nil {
+			return err
+		}
+	}
+
+	notMnt, err := ns.Mounter.IsLikelyNotMountPoint(mountPoint)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err = os.MkdirAll(mountPoint, os.FileMode(mountPermissions)); err != nil {
+				return err
+			}
+			notMnt = true
+		} else {
+			return err
+		}
+	}
+	if !notMnt {
+		log.DebugLog(ctx, "nfs: volume is already mounted to %s", mountPoint)
+
+		return nil
+	}
+
+	args := []string{
+		"-t", "nfs",
+		source,
+		mountPoint,
+	}
+
+	if len(mountOptions) > 0 {
+		args = append(append(args, "-o"), mountOptions...)
+	}
+
+	log.DefaultLog("nfs: mounting volumeID(%v) source(%s) targetPath(%s) mountflags(%v)",
+		volumeID, source, mountPoint, mountOptions)
+	if netNamespaceFilePath != "" {
+		_, stderr, err = util.ExecuteCommandWithNSEnter(
+			ctx, netNamespaceFilePath, "mount", args[:]...)
+	} else {
+		_, stderr, err = util.ExecCommand(ctx, "mount", args[:]...)
```

Comment:
consider using k8s.io/mount-utils for this, is has lots of error checking and is used elsewhere (RBD) already. Eventually we might be able to extend that package and include the `nsenter` wrapping there too.